Support color spaces (#2832)
sass/language: proposal/color-4-new-formats.md

# CSS Color Level 4, New Color Formats: Draft 1

*([Issue](https://github.com/sass/sass/issues/2831))*

This proposal adds several new color formats to Sass -- including `hwb()`,
`lab()`, and `lch()` as defined in [CSS Color Level 4][color-4].

[color-4]: https://www.w3.org/TR/css-color-4/

## Table of Contents

* [Background](#background)
* [Summary](#summary)
  * [Design Decisions](#design-decisions)
* [Semantics](#semantics)

## Background

> This section is non-normative.

The new modular syntax and name-spaced `sass:color` module allows us to provide
new CSS color formats before any browser implementations. The explicit purpose
of these Sass-module function will be providing pre-processed conversion to an
existing, well-supported color format.

While `hwb()` is defined in the sRGB space, with simple conversion, `lab()` &

> `lch()` colors are defined in a device-indipendant and *perceptually uniform*

Reviewer: ```suggestion
`lch()` colors are defined in a device-independant and *perceptually uniform*
```

> CIE space, much larger than the avialable sRGB gamut. That uniformity is useful

Reviewer: ```suggestion
CIE space, much larger than the available sRGB gamut. That uniformity is useful
```

out-of-gamut colors, so that they resolve to the nearest sRGB equivalent.
Colors that convert cleanly in-gamut will be un-affected. Since we are eagerly
resolving colors into sRGB space, while browsers can take a more nuanced
approach, it's important that we don't override users ability to use the CSS
functions once available.

With that in mind:
- None of these functions will be available on the global namespace.
- None of these functions will accept *special number string* or
  *special variable string* values that can only be resolved in CSS.

## Summary

> This section is non-normative.

This proposal defines Sassified versions of all the color functions in
[CSS Color Level 4][color-4]. Since the CIE color space defines the entire
gamut of visible color, much larger than the target sRGB gamut, out-of-range
color definitions will be clipped using a *relative-colorimetric* approach that
leaves in-gamut colors un-affected.

### Design Decisions

Conversions between color formats and color-spaces are well defined, and the
CSS specification provides code samples to help with implementation, but
handling out-of-gamut colors can be much more complex. There is no single

> "rendering intent" that works in every instance, and no well-defines algoythms

Reviewer: ```suggestion
"rendering intent" that works in every instance, and no well-defined algorithms
```
